### PR TITLE
feat: Draft release workflow

### PR DESCRIPTION
**Pull request by shrutiburman**
Fixes [DII-224](https://issues.corp.twilio.com/browse/DII-224), [DII-223](https://issues.corp.twilio.com/browse/DII-223), [DII-226](https://issues.corp.twilio.com/browse/DII-226)

This PR: 
Contains the entire draft release workflow from `main` branch. 
Contains changes from [PR](https://github.com/twilio/twilio-cli/pull/358)

Workflow runs: 
[Draft release](https://github.com/shrutiburman/twilio-cli/actions/runs/1771719527)
[Platform exec](https://github.com/shrutiburman/twilio-cli/actions/runs/1778037386)
[Platform exec](https://github.com/shrutiburman/twilio-cli/actions/runs/1773308818)

[Release link](https://github.com/shrutiburman/twilio-cli/releases/tag/3.0.0-draft.3)
### Checklist
- [x] I acknowledge that all my contributions will be made under the project's license
- [ ] I have made a material change to the repo (functionality, testing, spelling, grammar)
- [ ] I have read the [Contribution Guidelines](https://github.com/twilio/twilio-cli/blob/main/CONTRIBUTING.md) and my PR follows them
- [x] I have titled the PR appropriately
- [x] I have updated my branch with the main branch
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added the necessary documentation about the functionality in the appropriate .md file
- [ ] I have added inline documentation to the code I modified

If you have questions, please file a [support ticket](https://twilio.com/help/contact), or create a GitHub Issue in this repository.

### CODE REVIEW

File: .github/workflows/cli-test-homebrew.yml [hunk 1]
```diff
@@ -17,13 +17,21 @@ jobs:
         uses: actions/setup-node@v2
         with:
          node-version: ${{ matrix.node-version }}
+      - name: Check for cached twilio installation
```
Comment (shrutiburman) [inline .github/workflows/cli-test-homebrew.yml:20]: Reference: https://github.com/twilio/twilio-cli/pull/358#discussion_r777311205

File: .github/workflows/debian-draft-executable-release.yml [hunk 2]
```diff
@@ -28,20 +31,20 @@ jobs:
...
-        env:
-          SLACK_WEBHOOK: ${{ secrets.ALERT_SLACK_WEB_HOOK }}
-          SLACK_COLOR: "#ff3333"
-          SLACK_USERNAME: CLI Release Bot
-          SLACK_ICON_EMOJI: ":ship:"
-          SLACK_TITLE: "Twilio Cli"
-          SLACK_MESSAGE: 'Debian Executable Release Failed'
-          MSG_MINIMAL: actions url
\ No newline at end of file
+  # notify-complete-fail:
```
Comment (ravali-rimmalapudi) [inline .github/workflows/debian-draft-executable-release.yml:34]: Why we commented the notification part? Can we not specify draft release debian executable is failed?

File: .github/workflows/debian-draft-executable-release.yml [hunk 3]
```diff
@@ -19,6 +19,9 @@ jobs:
       - run: |
```
Comment (ravali-rimmalapudi) [inline .github/workflows/debian-draft-executable-release.yml:19]: Why we renamed only the debian file?

File: .github/workflows/draft-release.yml [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        with:
+          token: ${{ secrets.REPO_ACCESS_TOKEN }}
+      - name: Update Changelog
+        run: |
+          bash .github/scripts/commit-api-spec-change-log.sh "${{ github.event.inputs.change-log }}" ${{ github.event.inputs.version-type }}
+  release:
+    runs-on: ubuntu-latest
+    needs: [update-api-definitions-changelog]
+    outputs:
+      draft-tag-name: ${{ steps.semantic-release-draft.outputs.TAG_NAME }}
```
Comment (ravali-rimmalapudi) [inline .github/workflows/draft-release.yml:51]: Are we planning the clean the draft related code from main?

File: .github/workflows/draft-release.yml [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+        with:
+          ref: ${{steps.extract_branch.outputs.branch}}
+          check-name: 'Oclif Release'
+          repo-token: ${{ secrets.GITHUB_TOKEN }}
+          wait-interval: 10
+      - name: Run testing on release-draft
+        if: ${{ github.event.inputs.draft-release == 'true' }}
+        run: node .github/scripts/trigger-workflow.js
+        env:
+          WORKFLOW_NAME: '.github/workflows/cli-test-homebrew.yml'
```
Comment (ravali-rimmalapudi) [inline .github/workflows/draft-release.yml:128]: How are we ensuring homebrew is released before running homebrew tests?

File: .github/workflows/draft-release.yml (hunk 4 above)
Comment (shrutiburman) [reply .github/workflows/draft-release.yml:51]: Yes. [DII-309](https://issues.corp.twilio.com/browse/DII-309). This would also involve renaming some files in this branch so that merge with main is smooth without conflicts.

File: .github/workflows/debian-draft-executable-release.yml (hunk 2 above)
Comment (shrutiburman) [reply .github/workflows/debian-draft-executable-release.yml:34]: Since it's the draft release which would be monitored while running, instead of sending out multiple notifications I preferred sending one after the workflow had finished. Do you think we would need notifications, I can enable them all.

File: .github/workflows/debian-draft-executable-release.yml (hunk 3 above)
Comment (shrutiburman) [reply .github/workflows/debian-draft-executable-release.yml:19]: [platform-executables-acceptance-tests.yml](https://github.com/twilio/twilio-cli/pull/374/files#diff-2620ef6cacbc37efec19f000b7c02c86db2ee49aa70297c837ad874d04104b98)  and [windows-executable-release.yml](https://github.com/twilio/twilio-cli/pull/374/files#diff-2a58c3acab590ab41ebd782b3f7e1fec3a87b5edd4b7f408cd1a58e51ffc4db7) would also need renaming, I was going to rename them as a part of cleanup, but will get this done for better.

File: .github/workflows/draft-release.yml (hunk 5 above)
Comment (shrutiburman) [reply .github/workflows/draft-release.yml:128]: This action `lewagon/wait-on-check-action@v1.0.0` waits for the Oclif release to complete before triggering homebrew tests. 
 check-name: 'Oclif Release'

Comment (ravali-rimmalapudi) [reply .github/workflows/draft-release.yml:128]: But we should wait till hombrew workflow in homebrew repo right?

File: .github/workflows/debian-draft-executable-release.yml (hunk 2 above)
Comment (ravali-rimmalapudi) [reply .github/workflows/debian-draft-executable-release.yml:34]: If we monitor, removing make sense.

File: .github/workflows/draft-release.yml (hunk 5 above)
Comment (shrutiburman) [reply .github/workflows/draft-release.yml:128]: That's right. There's this default wait interval `wait-interval: 10`, but since there's no way to wait on another repo's workflow, I'd try to verify if this works fine in case of delayed releases.

**Review by ravali-rimmalapudi — APPROVED** [review]
LGTM:rocket:

File: .github/workflows/oclif-release.yml [hunk 6]
```diff
@@ -10,10 +10,10 @@ on:
         required: true
       formula:
           description: 'Formula to release in homebrew'
-          default: twilio
+          default: twiliodraft
```
Comment (shamantraghav) [inline .github/workflows/oclif-release.yml:13]: Why is this changed to twiliodraft? Won't this affect the regular release as the main formula won't get updated?

File: .github/workflows/cli-test-npm.yml [hunk 7]
```diff
@@ -17,22 +17,40 @@ jobs:
...
+        uses: actions/checkout@v2
+      - name: Use Node.js ${{ matrix.node-version }}
+        uses: actions/setup-node@v2
+        with:
+         node-version: ${{ matrix.node-version }}
+      - name: Check for cached twilio installation 
+        run: | 
+         if ! command -v twilio > /dev/null; then twilio() { printf '%sCLI is not installed, all good. '; } else echo "CLI is installed already, Abort."; exit 1; fi
+         twilio --version
+      - name: Check Out Repo
```
Comment (shamantraghav) [inline .github/workflows/cli-test-npm.yml:30]: Why are we checking out the repo again here? We do this in Line 20 as well

File: .github/workflows/oclif-release.yml (hunk 6 above)
Comment (shrutiburman) [reply .github/workflows/oclif-release.yml:13]: This `oclif-release.yml` file will only be in this draft-release branch, similar to what we did for release-feature branch. We would be merging main branch to this not the other way around.

File: .github/workflows/cli-test-npm.yml (hunk 7 above)
Comment (shrutiburman) [reply .github/workflows/cli-test-npm.yml:30]: That can be removed, it's only needed once below. Taking it as a nit, I'll clean it up in the upcoming PR.
